File: app/src/main/java/forpdateam/ru/forpda/client/ProgressRequestBody.java
package forpdateam.ru.forpda.client;

import android.support.annotation.NonNull;

import java.io.IOException;

import forpdateam.ru.forpda.model.data.remote.IWebClient;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Okio;
import okio.Sink;

/**
 * Created by radiationx on 10.11.17.
 */
public class ProgressRequestBody extends RequestBody {
    private RequestBody mDelegate;
    private IWebClient.ProgressListener mListener;

    ProgressRequestBody(RequestBody delegate, IWebClient.ProgressListener listener) {
        mDelegate = delegate;
        mListener = listener;
    }

    @Override
    public MediaType contentType() {
        return mDelegate.contentType();
    }

    @Override
    public long contentLength() {
        try {
            return mDelegate.contentLength();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    @Override
    public void writeTo(@NonNull BufferedSink sink) throws IOException {
        CountingSink mCountingSink = new CountingSink(sink);
        BufferedSink bufferedSink = Okio.buffer(mCountingSink);
        mDelegate.writeTo(bufferedSink);
        bufferedSink.flush();
    }

    private final class CountingSink extends ForwardingSink {
        private long bytesWritten = 0;

        CountingSink(Sink delegate) {
            super(delegate);
        }

        @Override
        public void write(@NonNull Buffer source, long byteCount) throws IOException {
            super.write(source, byteCount);
            bytesWritten += byteCount;
            mListener.onProgress((int) (100F * bytesWritten / contentLength()));
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/client/ProgressRequestBody.kt
package forpdateam.ru.forpda.client

import forpdateam.ru.forpda.model.data.remote.IWebClient
import okhttp3.MediaType
import okhttp3.RequestBody
import okio.Buffer
import okio.BufferedSink
import okio.ForwardingSink
import okio.Sink
import okio.buffer

/**
 * Created by radiationx on 10.11.17.
 */
class ProgressRequestBody(
    private val requestBody: RequestBody,
    private val listener: IWebClient.ProgressListener
) : RequestBody() {

    override fun contentType(): MediaType? {
        return requestBody.contentType()
    }

    override fun contentLength(): Long {
        return requestBody.contentLength()
    }

    override fun writeTo(sink: BufferedSink) {
        val countingSink = CountingSink(sink)
        val bufferedSink = countingSink.buffer()
        requestBody.writeTo(bufferedSink)
        bufferedSink.flush()
    }

    private inner class CountingSink(sink: Sink) : ForwardingSink(sink) {
        private var bytesWritten: Long = 0

        override fun write(source: Buffer, byteCount: Long) {
            super.write(source, byteCount)
            bytesWritten += byteCount
            listener.onProgress((100f * bytesWritten / contentLength()).toInt())
        }
    }
}
